sha2/src/lib.rs

```diff
@@ -13,23 +13,37 @@
 #[cfg(not(any(target_arch = "x86_64", target_arch = "x86", target_arch = "aarch64")))]
 compile_error!("crate can only be used on x86, x86-64 and aarch64 architectures");
 
+cpufeatures::new!(cpuid_avx2, "avx2");
+
 #[link(name = "sha256", kind = "static")]
+#[allow(dead_code)]
 extern "C" {
     fn sha256_compress(state: &mut [u32; 8], block: &[u8; 64]);
+    fn sha256_transform_rorx(state: &mut [u32; 8], block: *const [u8; 64], num_blocks: u64);
 }
 
 /// Safe wrapper around assembly implementation of SHA256 compression function
+///
 #[inline]
 pub fn compress256(state: &mut [u32; 8], blocks: &[[u8; 64]]) {
-    for block in blocks {
-        unsafe { sha256_compress(state, block) }
+    let token: cpuid_avx2::InitToken = cpuid_avx2::init();
+
+    if token.get() {
+        if !blocks.is_empty() {
+            unsafe { sha256_transform_rorx(state, blocks.as_ptr(), blocks.len() as u64) }
+        }
+    } else {
+        for block in blocks {
+            unsafe { sha256_compress(state, block) }
+        }
     }
 }
 
 #[cfg(not(target_arch = "aarch64"))]
 #[link(name = "sha512", kind = "static")]
 extern "C" {
     fn sha512_compress(state: &mut [u64; 8], block: &[u8; 128]);
+    fn sha512_transform_rorx(state: &mut [u64; 8], block: *const [u8; 128], num_blocks: u64);
```

Comment:
Are you sure that `num_blocks` should be `u64` here? In the assembly comment this function has the following signature:
```C
void sha512_transform_rorx(sha512_state *state, const u8 *data, int blocks)
```
If I am reading the assembly correctly, this argument is passed in register, so it probably should be `usize`.